File: src/DataStructure/DesignTwitter.java
package DataStructure;

import java.util.*;

// https://leetcode.com/problems/design-twitter/
public class DesignTwitter {
    private HashMap<Integer, HashSet<Integer>> follows; // userId as key
    private HashMap<Integer, LinkedList<Tweet>> tweets; // userId as key
    private int timestamp;

    private class Tweet implements Comparable<Tweet> {
        private int tweetId;
        private int ts;
        private int userId;

        public Tweet(int tweetId, int userId, int timestamp) {
            this.tweetId = tweetId;
            this.ts = timestamp;
            this.userId = userId;
        }

        @Override
        public int compareTo(Tweet t2) {
            return t2.ts - this.ts; // maxHeap
        }
    }

    /**
     * Initialize your data structure here.
     */
    public DesignTwitter() {
        follows = new HashMap<>();
        tweets = new HashMap<>();
        timestamp = 0;
    }

    /**
     * Compose a new tweet.
     */
    public void postTweet(int userId, int tweetId) {
        if (!tweets.containsKey(userId)) {
            tweets.put(userId, new LinkedList<>());
        }
        tweets.get(userId).add(0, new Tweet(tweetId, userId, timestamp++)); // put to index=0, most recent

        if (!follows.containsKey(userId)) { // follow self
            follows.put(userId, new HashSet<>());
            follows.get(userId).add(userId);
        }
    }

    /**
     * Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by
     * users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent.
     */
    public List<Integer> getNewsFeed(int userId) {
        List<Integer> feed = new LinkedList<>();
        PriorityQueue<Tweet> q = new PriorityQueue<>();
        if (!follows.containsKey(userId)) {
            return feed;
        }
        HashSet<Integer> followed = follows.get(userId);
        HashMap<Integer, Integer> count = new HashMap<>(); // track the current userId's position - where to fetch the next
        for (Integer followedUserId : followed) {
            if (tweets.containsKey(followedUserId)) {
                Tweet t = tweets.get(followedUserId).get(0);
                q.add(t);
                count.put(t.userId, 1);
            }
        }
        while (q.size() > 0 && feed.size() < 10) {
            Tweet t = q.poll();
            feed.add(t.tweetId);
            int next = count.get(t.userId); // refill the heap - greedy
            count.put(t.userId, next + 1);
            if (next < tweets.get(t.userId).size()) { // if there is still more tweets from this userId
                q.add(tweets.get(t.userId).get(next));
            }
        }
        return feed;
    }

    /**
     * Follower follows a followee. If the operation is invalid, it should be a no-op.
     */
    public void follow(int followerId, int followeeId) {
        if (!follows.containsKey(followerId)) {
            follows.put(followerId, new HashSet<>());
            follows.get(followerId).add(followerId);
        }
        follows.get(followerId).add(followeeId);
    }

    /**
     * Follower unfollows a followee. If the operation is invalid, it should be a no-op.
     */
    public void unfollow(int followerId, int followeeId) {
        if (followerId == followeeId) {
            return;
        }
        if (follows.containsKey(followerId)) {
            follows.get(followerId).remove(followeeId);
        }
    }
}

File: py/essential/DesignTwitter.py
# https://leetcode.com/problems/design-twitter/description/?envType=problem-list-v2&envId=plakya4j
# Neet

import collections
from typing import List


class Twitter:
    def __init__(self):
        self.usersPosts=collections.defaultdict(list)
        self.time=1
        self.followersToFollowees=collections.defaultdict(set)

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.usersPosts[userId].append((self.time, tweetId))
        self.time+=1
        
    def getNewsFeed(self, userId: int) -> List[int]:
        allTweets=self.usersPosts[userId].copy()
        for followeeId in self.followersToFollowees[userId]:
            if followeeId in self.usersPosts:
                allTweets+=self.usersPosts[followeeId]
        allTweets.sort(reverse=True, key=lambda x:x[0])
        return [tweetId for _, tweetId in allTweets[:10]]

    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId!=followeeId:
            self.followersToFollowees[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followeeId in self.followersToFollowees[followerId]:
            self.followersToFollowees[followerId].remove(followeeId)

s=Twitter()
s.postTweet(1, 5)
s.postTweet(1, 6)
s.postTweet(1, 7)
s.postTweet(1, 8)
s.follow(1, 2)
s.postTweet(2, 10)
s.postTweet(2, 11)
print(s.getNewsFeed(1))
print(s.getNewsFeed(2))
